### PR TITLE
Add Improve & Update Typechain Polkadot

### PR DESCRIPTION
**Pull request by WookashWackomy**
#### Project Abstract

A reliable, developer friendly typegen tool is crucial for the developer ecosystem.
This application aims to improve Typechain Polkadot, extending its capabilities & reworking parts of it that make it cumbersome to use. 

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] Payment details have been provided (Polkadot AssetHub (DOT, USDC & USDT) address in the application and bank details via email, if applicable).
- [x] I understand that 30% of each milestone will be paid in vested DOT, to the Polkadot address listed in the application.
- [x] I am aware that, in order to receive a grant, I (and the entity I represent) have to successfully complete a KYC/KYB check.
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

### CODE REVIEW

**Comment by WookashWackomy** [comment]
I have read and hereby sign the Contributor License Agreement.

**Review by keeganquigley — CHANGES_REQUESTED** [review]
Hi @WookashWackomy thanks for the application. Usually we require all the mandatory deliverables 0a - 0e. that are present in the [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md). Could you please add these to each milestone and move the current deliverables underneath starting with 1., 2., etc.

An article is only mandatory for the last milestone, and if we can run everything with pnpm or yarn it's probably fine to remove Docker. Thanks!

**Comment by WookashWackomy** [comment]
> Hi @WookashWackomy thanks for the application. Usually we require all the mandatory deliverables 0a - 0e. that are present in the [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md). Could you please add these to each milestone and move the current deliverables underneath starting with 1., 2., etc.
> 
> An article is only mandatory for the last milestone, and if we can run everything with pnpm or yarn it's probably fine to remove Docker. Thanks!

Hello, sure. Just updated the proposal with your suggestions. I've left the Docker but marked as optional - perhaps it would be handy to provide a Dockerfile with a strictly set runtime & all of the deps just for the sake of testing the project without the need to use local environment. Though I'm not 100% sure if it makes sense in a real life scenario. That being said - it's there and we'll try to make it in a way to be useful.  Thanks for the feedback!

File: applications/typechain_revived.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+#### Remove ReturnNumber
+The ReturnNumber wrapper class is unnecessary and makes the code harder to read and use. Introduces runtime bugs, false assumptions and redundant additional complexity on top of polkadot.js. We will remove it and replace it with BN.
+
+#### Update to ink! 5
+Update Typechain to work with ink! 5 - specifically to handle the way events are being presented in contracts ABI.
+
+#### Simplify & enhance usage of deployment classes
+The current design of Typechain generated code induces runtime errors due to improper handling of ABI. Our goal is to make the code runtime-safe and more intuitive.
+    Additinally we will add a possibility to upload code without instantiating the contract. (Corresponding issue)[https://github.com/Brushfam/typechain-polkadot/issues/106]
```
Comment (semuelle) [inline applications/typechain_revived.md:32]:
```suggestion
    Additinally we will add a possibility to upload code without instantiating the contract. [Corresponding issue](https://github.com/Brushfam/typechain-polkadot/issues/106)
```

**Review by semuelle — COMMENTED** [review]
Thanks for the application. What's the status of the Brushfam repo, is Brushfam still involved in Typechain's maintenance? My main concern for such a grant would be that we need to make sure that the work is merged into the main repo, instead of you having to maintain a fork and thus possibly splitting the community.

**Comment by WookashWackomy** [comment]
> Thanks for the application. What's the status of the Brushfam repo, is Brushfam still involved in Typechain's maintenance? My main concern for such a grant would be that we need to make sure that the work is merged into the main repo, instead of you having to maintain a fork and thus possibly splitting the community.

Brushfam has not been maintaing the project for a long time and did clearly state that if we, community want to make any updates/changes we have to create a fork :/
There is a fork, maintained by prosopo team (https://github.com/prosopo/typechain-polkadot), however:
1) it contains no functional changes or bug fixes (besides 1 typescript only, non runtime related fix & merging one external PR ) - just updating dependencies
2) We have contacted the team and they don't plan on making any adjustments/updates to typechain - definitely not during next few months. That, in our opinion, emphasizes the justification to undertake the actions mentioned in the proposal. Especially given the short timeline of the proposed actions (which we are very confident of being able to deliver due to our familiarity with the project overall & a good understanding of project's code & polkadotjs overall).

Regarding hosting:
We really regret the state typechain-polkadot has found itself in and do not care that much where it's hosted as long as it's in a good, usable, user friendly state and the community has a platform to submit & make changes to the project.

We've been debating about hosting it in an appropriate place for such project - https://github.com/inkdevhub/ and creating a wide group of maitainers from different teams/projects. That would make possible to act quickly at least when a community member creates a PR - which was difficult to achieve for for a long time and forced people to fork the project.
However we could submit the work on a branch in prosopo's repo to have it merged in out there.

(Offtopic note - we've been thinking about hosting [Pendzl](https://github.com/pendzl/Pendzl) on inkdevhub as well and will attempt to do it soon)

File: applications/typechain_revived.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
+# Improve & Update Typechain Polkadot
+
+- **Team Name:** C Forge
+- **Payment Address:** 15VHeCpx5SUbRsbiZAfYtERxXPNKNZfc9bNBnGGZ4SvaT2ct (USDT)
```
Comment (semuelle) [inline applications/typechain_revived.md:4]: Please indicate whether you want this address to be used for DOT payments on Polkadot as well, or a separate address.

**Review by semuelle — CHANGES_REQUESTED** [review]
inkdevhub might be a good place. Once the ink! team has reestablished itself, they might have an opinion on this as well.

File: applications/typechain_revived.md (hunk 2 above)
Comment (WookashWackomy) [reply applications/typechain_revived.md]: Hi, sorry, I thought it was implied by PRs description checkbox. I did put a notice in the PR itself now.

Comment (semuelle) [reply applications/typechain_revived.md]: Ah, sorry, that's actually outdated as we transfer DOT on Polkadot directly, not AssetHub.

**Comment by WookashWackomy** [comment]
Hello @semuelle. Could you share an update on the state of the approval process of this proposal?

**Review by semuelle — APPROVED** [review]
Hi @WookashWackomy, sorry for the delay. I am happy to support the project and will share it with the rest of the committee. They will probably have comments and questions in the coming days.

**Comment by keeganquigley** [comment]
Hi @WookashWackomy can you please also complete [KYB verification](https://in.sumsub.com/idensic/l/#/Qm32HqN12b7PrrOB)? This is required for all grantees, thank you!

**Comment by WookashWackomy** [comment]
> Thanks @WookashWackomy based on your previous work I'm also happy to approve. But curious what steps you will take afterwards to gain more traction with this tool?

Thanks @keeganquigley for the approve. Let me somehow reiterate here - the goal of so called "typechain" is to be an important piece & facilitator in areas such as building, scripting, testing etc. Therefore a great, and honestly the best way to get traction is to prove it is such building block indeed - by providing examples of usage in real life projects. 
Besides us having plans to build/refine and then release to public a number of projects/scripts after completing the work mentioned in the proposal....we firmly believe that this set of typechain libraries will start getting traction itself by being the most compelling and easy to pick up tooling.

**Review by Noc2 — APPROVED** [review]

**Pull request merged**